syntax = "proto3";

package protos;
option go_package = "orch-go/gen/protos;protos";
option csharp_namespace = "Core";

import "customer.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";
import "custom_types.proto";
import "account.proto";
import "card.proto";
import "transaction.proto";
import "loan.proto";
import "deposit.proto";
import "login_log.proto";


service CustomerAnalyticService {
  rpc GetCustomersByBalanceRange (GetCustomersByBalanceRangeRequest) returns (GetAllCustomersResponse);
  rpc GetCustomersByTransactionQuantityRange (GetCustomersByTransactionQuantityRangeRequest) returns (GetAllCustomersResponse);
  rpc GetCustomerLifeTime (GetCustomerLifeTimeRequest) returns (GetCustomerLifeTimeResponse);
  rpc GetCustomersLifeTime (google.protobuf.Empty) returns (GetCustomersLifeTimeResponse);
  rpc GetInactiveCustomers (google.protobuf.Empty) returns (GetAllCustomersResponse);
  rpc GetCountByBalanceRange (GetCustomersByBalanceRangeRequest) returns (CountResponse);
  rpc GetCountByTransactionQuantityRange (GetCustomersByTransactionQuantityRangeRequest) returns (CountResponse);
  rpc GetCountInactiveCustomers (google.protobuf.Empty) returns (CountResponse);
  rpc GetCustomerDetails (GetCustomerDetailsRequest) returns (CustomerDetailsResponse);
}


message GetCustomersByBalanceRangeRequest {
  string min_balance = 1;
  string max_balance = 2;
}

message GetCustomersByTransactionQuantityRangeRequest{
  int32 min_quantity = 1;
  int32 max_quantity = 2;
}

message GetCustomerLifeTimeRequest {
  int32 customer_id = 1;
}

message GetCustomerLifeTimeResponse {
  int32 customer_id = 1;
  google.protobuf.Timestamp life_time = 2;
}

message GetCustomersLifeTimeResponse {
  repeated GetCustomerLifeTimeResponse customer_life_time = 1;
}

message GetCustomerDetailsRequest {
  int32 customer_id = 1;
  optional int32 transactions_page_n = 2;
  optional int32 transactions_page_size = 3;
  optional int32 activity_page_n = 4;
  optional int32 activity_page_size = 5;
}

message CustomerDetailsResponse {
  CustomerModel customer = 1;
  repeated AccountModel accounts = 2;
  repeated CardModel cards = 3;
  repeated TransactionModel transactions = 4;
  repeated LoanModel loans = 5;
  repeated DepositModel deposits = 6;
  repeated LoginLogModel activity = 7;
  
  optional int32 transactions_total_count = 8;
  optional int32 activity_total_count = 9;
}
